explorer/interpreter/impl_scope.cpp

```diff
@@ -103,7 +103,33 @@ auto ImplScope::Resolve(Nonnull<const Value*> constraint_type,
                            source_loc, type_checker));
       witnesses.push_back(result);
     }
-    // TODO: Check satisfaction of same-type constraints.
+
+    // Check that all equality constraints are satisfied in this scope.
+    if (llvm::ArrayRef<EqualityConstraint> equals =
+            constraint->equality_constraints();
+        !equals.empty()) {
+      std::optional<Nonnull<const Witness*>> witness;
+      if (constraint->self_binding()->impl_binding()) {
+        witness = type_checker.MakeConstraintWitness(*constraint, witnesses,
+                                                     source_loc);
+      }
+      Bindings local_bindings = bindings;
+      local_bindings.Add(constraint->self_binding(), impl_type, witness);
+      SingleStepEqualityContext equality_ctx(&type_checker, this);
+      for (auto& equal : equals) {
+        auto it = equal.values.begin();
+        Nonnull<const Value*> first =
+            type_checker.Substitute(local_bindings, *it++);
+        for (; it != equal.values.end(); ++it) {
+          Nonnull<const Value*> current =
+              type_checker.Substitute(local_bindings, *it);
+          if (!ValueEqual(first, current, &equality_ctx)) {
+            return ProgramError(source_loc) << "could not determine that "
+                                            << *first << " == " << *current;
```

Comment:
Wording improved. I'm avoiding saying "which isn't provable" because there are cases where such things are provable but we nonetheless can't discover the proof for ourselves. I've also added a test with an example of that.